### PR TITLE
DolphinQt: Add dark theme

### PR DESCRIPTION
**Pull request by FearlessTobi**
This PR adds a dark theme to the DolphinQt frontend.
It's based on QDarkStyle with some minor changes to make it look a bit nicer in Dolphin.

Also Dolphin will now register the QResources for stylesheets properly, if an .rcc file is found in the folder of the UserStyle.
This makes it possible to distribute a Dolphin UserStyle in one single folder, instead of requiring the user to put their stylesheet and their icons in different places.
To generate the .rcc file from a .qrc file, run `rcc -binary style.qrc -o style.rcc` with Qt installed.

Updated screenshot of the changes:
grafik

### CODE REVIEW

**Comment by Simonx22** [comment]
@dolphin-emu-bot rebuild

**Comment by FearlessTobi** [comment]
> It looks good. Somewhat surprised it requires(?) pngs and can't all be done in code, but whatever.

This is just how QDarkStyleSheet does it... it probably can be done in code, but I hope you understand that I'm more willing to just drop in that theme instead of extensively rewriting it first.

> It would be nicer if it were just the dark windows theme (of whichever windows/OS version you're running, but I wouldn't be surprised if that's not possible).

I sadly do not know of a way to do this in QT on Windows.

> is it possible to integrate the rcc step into the build system? or is it preferable not to?

Yes, CMake even has a neat AUTORCC variable for this. But it seems like the Windows build is not even using CMake(?).
This is my first time contributing to Dolphin and I haven't fully figured out yet how your build system works.

> my personal preference would be to have the style not be directly configurable and just use whatever the OS is set to 

I agree with the sentiment but the combobox has already been there before. It was just hidden behind the "Use Custom Userstyle" option. With this PR if a userstyle is added, it will just be shown in the Theme combobox as well.

**Comment by MayImilae** [comment]
This really is something that should be taken care of by the operating system but ~Windows problems~. Shipping a dark theme is fine.

Regarding this specific theme, it's has a bit of a colour cast to it. For example, the background of the settings window is 49, 54, 49 RGB. Why is it green? IMO, I'd prefer it just be a perfectly colourless 49, 49, 49.

**Comment by Starsam80** [comment]
I don't know if opinions have changed in ~5 years, but the decision to have a separate checkbox + dropdown was very intentional: https://github.com/dolphin-emu/dolphin/pull/6768#issuecomment-387185124

**Comment by MayImilae** [comment]
Yea this is a little different as this isn't a custom user style, it's shipping with Dolphin.

Speaking of which, how do those still work with this? Have you tested it?

**Comment by FearlessTobi** [comment]
From my limited testing, they should work fine and are added to the Theme combobox as well.

dolphin2

**Comment by FearlessTobi** [comment]
> Regarding this specific theme, it's has a bit of a colour cast to it. For example, the background of the settings window is 49, 54, 49 RGB. Why is it green? IMO, I'd prefer it just be a perfectly colourless 49, 49, 49.

Alright I adjusted the colors. Personally I think it looks a bit worse now, but it doesn't matter much.

If there's anything else I can do to move this PR further along, just let me know.

**Comment by Simonx22** [comment]
@dolphin-emu-bot rebuild

**Comment by FearlessTobi** [comment]
Finally got around to updating this PR.
Now the dark mode of Dolphin gets selected automatically if the dark theme is enabled in Windows.

**Comment by Simonx22** [comment]
@dolphin-emu-bot rebuild

**Comment by AdmiralCurtiss** [comment]
@shuffle2 You said you could do the `rcc` call in the VS project files, do you want to implement that?

**Comment by AdmiralCurtiss** [comment]
Okay I'm playing around with this now and this seems totally doable to embed. https://github.com/AdmiralCurtiss/dolphin/commit/0618c671bf1224c2cc5729c98cbeaac605532618 is the CMake implementation, I'll see if I can get the Visual Studio one working...

**Comment by AdmiralCurtiss** [comment]
Ha alright, in that case I'll just leave it to you, you definitely know VS's build system better than me.

File: Source/Core/DolphinQt/Settings.cpp [hunk 1]
```diff
@@ -126,19 +129,45 @@ QString Settings::GetCurrentUserStyle() const
...
   {
-    // Load custom user stylesheet
-    QDir directory = QDir(QString::fromStdString(File::GetUserPath(D_STYLES_IDX)));
-    QFile stylesheet(directory.filePath(stylesheet_name));
+    const auto& stylesheet_path = File::GetStylesDir(stylesheet_name.toStdString());
+    if (stylesheet_path)
+    {
+      // Load stylesheet
+      QDir directory = QDir(QString::fromStdString(stylesheet_path.value()));
+      QFile stylesheet(directory.filePath(QStringLiteral("style.qss")));
```
Comment (AdmiralCurtiss) [inline Source/Core/DolphinQt/Settings.cpp:161]: Doesn't this break stylesheets that are not called `style.qss`...?

File: Source/Core/DolphinQt/Settings.cpp [hunk 2]
```diff
@@ -126,19 +129,45 @@ QString Settings::GetCurrentUserStyle() const
...
+      QString resource_path = directory.filePath(QStringLiteral("style.rcc"));
+
+      if (stylesheet.open(QFile::ReadOnly))
+        stylesheet_contents = QString::fromUtf8(stylesheet.readAll().data());
 
-    if (stylesheet.open(QFile::ReadOnly))
-      stylesheet_contents = QString::fromUtf8(stylesheet.readAll().data());
+      // Register QResource file if it exists
+      if (QFile(resource_path).exists())
+        QResource::registerResource(resource_path);
```
Comment (AdmiralCurtiss) [inline Source/Core/DolphinQt/Settings.cpp:169]: This seems really janky. If you want to support loading .rcc files, you should load the entire stylesheet from it, not just the images. You should also unregister the resource when the style is switched again, as you can't guarantee multiple styles won't share the same image filenames. The `mapRoot` parameter to `registerResource()` should also probably be set to something that is guaranteed non-conflicting with other resources.

**Review by AdmiralCurtiss — CHANGES_REQUESTED** [review]
After trying it out, I think we should definitely embed the theme instead of loading it from the Data dir. I also have issues with your QResource handling.

File: Source/Core/DolphinQt/Settings.cpp (hunk 2 above)
Comment (AdmiralCurtiss) [reply Source/Core/DolphinQt/Settings.cpp:169]: In fact, this seems complex enough to handle correctly that I'd prefer the stylesheet loading logic changes in a separate PR.

**Comment by FearlessTobi** [comment]
Alright, I will update the PR after you share a working commit of AutoRCC in the MSVC build system.

File: Source/Core/DolphinQt/Settings.cpp (hunk 2 above)
Comment (FearlessTobi) [reply Source/Core/DolphinQt/Settings.cpp]: I agree. The changes have been removed from this PR.

File: Source/Core/DolphinQt/Settings.cpp (hunk 1 above)
Comment (FearlessTobi) [reply Source/Core/DolphinQt/Settings.cpp]: fixed.

**Comment by AdmiralCurtiss** [comment]
I'm not sure if you've seen this but we've decided to wait until the Qt6.5 version update (which should be in a week or two), since that apparently is adding automatic dark theme support for Windows. If you want to add your other changes you should probably open a separate PR to discuss them there.

https://github.com/dolphin-emu/dolphin/pull/11672#issuecomment-1475420467

**Comment by Moonlacer** [comment]
What's the current state of this PR? It seems there was some similar work done on #11672, but nothing has been done on that one in roughly three weeks

**Comment by FearlessTobi** [comment]
Apart from the minor conflict I need to fix, this PR would be pretty much ready to go.
There was also https://github.com/dolphin-emu/dolphin/pull/11789 open as an alternative solution, but since that has been closed again, it would be great if I could get a re-review here.

**Comment by Moonlacer** [comment]
I know I'm looking forward to a dark theme, I would love to see this merged to save my eyes without having to use a third-party tool. It honestly really surprised me that there weren't any themes to select from still when I recently reinstalled Dolphin

**Comment by FearlessTobi** [comment]
Rebased.

> I know I'm looking forward to a dark theme, I would love to see this merged to save my eyes without having to use a third-party tool. It honestly really surprised me that there weren't any themes to select from still when I recently reinstalled Dolphin

I agree, Dolphin as an emulator is pretty much perfect, but it's still missing this basic feature.
Not getting flashbanged every time you open it would be very nice :p

**Comment by AdmiralCurtiss** [comment]
Yeah, you're right, we should do this. I think this needs to be done in two steps: First we decide which theme we actually want to use (since there's several options at this point), and then we do new PR with a clean implementation of that.

For comparison, this is a shot of our current 'Light' theme that we use on Windows: [Current Light Style](https://github.com/dolphin-emu/dolphin/assets/4522237/701c1586-1358-4710-9d72-8c8e900e5a2f)


With that, our options are: (Note that I've avoided commenting on the aesthetics, this is just in terms of facts.)

- QtFusion. -- [Screenshot (Light)](https://github.com/dolphin-emu/dolphin/assets/4522237/417f430e-5c26-4a91-bc76-d0bdfa5876bd), [Screenshot (Dark)](https://github.com/dolphin-emu/dolphin/assets/4522237/aa70dac7-1e04-4515-8301-8241d4f0866a)
  - Advantage: Provided by Qt. No extra work has to be done on our end.
  - Disadvantage: This is the only option that actually affects users of the Light theme as well.

- #11446 by FearlessTobi based on QDarkStyle (supposedly, I can't actually find a matching source for this). -- [Screenshot](https://github.com/dolphin-emu/dolphin/assets/4522237/fe3f11dc-2969-48f3-83d3-a9e7a8c536db)
  - Advantage: Done, we can use this as-is.
  - Disadvantage: Need to ship a license for the theme, which we need to figure out how/where to do that.
  - Disadvantage: Source unclear. This can hopefully be resolved easily.

- [Dolphin Dark Mode](https://github.com/EndangeredNayla/Dolphin-Dark-Mode) by EndangeredNayla based on [QDarkStyleSheet](https://github.com/ColinDuquesnoy/QDarkStyleSheet/blob/master/qdarkstyle/dark/darkstyle.qss). -- [Screenshot](https://github.com/dolphin-emu/dolphin/assets/4522237/8283015e-bc49-4410-9555-8b484d701978)
  - Advantage: Done, we can use this as-is.
  - Disadvantage: Need to ship a license for the theme, which we need to figure out how/where to do that.

- #11672 by AdmiralCurtiss. -- [Screenshot](https://github.com/dolphin-emu/dolphin/assets/4522237/6b7f4718-dc4a-4695-85d5-be5f66d0eccd)
  - Advantage: Made from scratch, no extra license needed.
  - Disadvantage: Not fully complete, still needs some tweaking to actually look correct.

(technically anyone else could come along with another theme, but that's what's been suggested or PR'd so far)

I'll poke some other devs and ask about what we'd prefer to do here.

**Comment by FearlessTobi** [comment]
>     QtFusion. -- [Screenshot (Light)](https://github.com/dolphin-emu/dolphin/assets/4522237/417f430e-5c26-4a91-bc76-d0bdfa5876bd), [Screenshot (Dark)](https://github.com/dolphin-emu/dolphin/assets/4522237/aa70dac7-1e04-4515-8301-8241d4f0866a)
>         Advantage: Provided by Qt. No extra work has to be done on our end.
>         Disadvantage: This is the only option that actually affects users of the Light theme as well.
I honestly find this to look very dated, and as you already said, it makes the Light theme look worse as well.


> https://github.com/dolphin-emu/dolphin/pull/11446 by FearlessTobi based on QDarkStyle (supposedly, I can't actually find a matching source for this). -- [Screenshot](https://github.com/dolphin-emu/dolphin/assets/4522237/fe3f11dc-2969-48f3-83d3-a9e7a8c536db)
> 
>     Advantage: Done, we can use this as-is.
>     Disadvantage: Need to ship a license for the theme, which we need to figure out how/where to do that.
>     Disadvantage: Source unclear. This can hopefully be resolved easily.
I'm not sure if you missed it, but the MIT license is already being shipped in the QDarkStyle directory with this Pull Request.
If this is not good enough for you, we can probably add the SPDX header in the .qss file as well.

As for the source, I think I more or less directly copied this over from yuzu's dark style, as I was already very familiar with it.
It seems to be originally from this commit https://github.com/yuzu-emu/yuzu/commit/1355f0f39e0e66aa6ed7e8a9a2ac122b57a62255, which is in turn based on https://github.com/ColinDuquesnoy/QDarkStyleSheet and received multiple updates from yuzu contributors over the years.


> [Dolphin Dark Mode](https://github.com/EndangeredNayla/Dolphin-Dark-Mode) by EndangeredNayla based on [QDarkStyleSheet](https://github.com/ColinDuquesnoy/QDarkStyleSheet/blob/master/qdarkstyle/dark/darkstyle.qss?rgh-link-date=2023-07-29T15%3A38%3A49Z). -- [Screenshot](https://github.com/dolphin-emu/dolphin/assets/4522237/8283015e-bc49-4410-9555-8b484d701978)
> 
>     Advantage: Done, we can use this as-is.
>     Disadvantage: Need to ship a license for the theme, which we need to figure out how/where to do that.
Same with the licensing applies here. Since Dolphin is GPL and the dark theme is licensed under MIT, I don't see any issues with shipping it.

> https://github.com/dolphin-emu/dolphin/pull/11672 by AdmiralCurtiss. -- [Screenshot](https://github.com/dolphin-emu/dolphin/assets/4522237/6b7f4718-dc4a-4695-85d5-be5f66d0eccd)
> 
>     Advantage: Made from scratch, no extra license needed.
>     Disadvantage: Not fully complete, still needs some tweaking to actually look correct.
As you said yourself, this is still incomplete and would need considerable work to be worth considering.

**Comment by AdmiralCurtiss** [comment]
In case you haven't been following the discussion, we've decided to go with #12080. This means your stylesheet/resource loading logic changes are now orphaned here, if you care about them you can open another PR with them. (though I still find the idea of loading arbitrary user-provided resource files to be pretty sketchy, if I'm being honest)

**Pull request closed without merging**